server/ec2spotmanager/cron.py

```diff
@@ -107,33 +107,30 @@ def check_instance_pools():
 
 @app.task
 def update_spot_prices():
+    from .models import PoolConfiguration
+
     """Periodically refresh spot price history and store it in redis to be consumed when spot instances are created.
 
     Prices are stored in redis, with keys like:
         'ec2spot:price:{instance-type}'
     and values as JSON objects {region: {az: [prices]}} like:
         '{"us-east-1": {"us-east-1a": [0.08000, ...]}}'
     """
-    from .common.prices import get_spot_prices
-    from .models import PoolConfiguration
 
     regions = set()
     for cfg in PoolConfiguration.objects.all():
-        if cfg.ec2_allowed_regions:
-            regions |= set(json.loads(cfg.ec2_allowed_regions))
-
-    now = timezone.now()
-    expires = now + datetime.timedelta(hours=3)  # how long this data is valid (if not replaced)
-
-    prices = get_spot_prices(regions,
-                             getattr(settings, 'AWS_ACCESS_KEY_ID', None),
-                             getattr(settings, 'AWS_SECRET_ACCESS_KEY', None))
-
-    # use pipeline() so everything is in 1 transaction
-    cache = redis.StrictRedis(host=settings.REDIS_HOST, port=settings.REDIS_PORT, db=settings.REDIS_DB).pipeline()
-    for instance_type in prices:
-        key = 'ec2spot:price:' + instance_type
-        cache.delete(key)
-        cache.set(key, json.dumps(prices[instance_type], separators=(',', ':')))
-        cache.expireat(key, expires)
-    cache.execute()  # commit to redis
+        if cfg.allowed_regions:
+            regions |= set(json.loads(cfg.allowed_regions))
+
+        now = timezone.now()
+        expires = now + datetime.timedelta(hours=12)  # how long this data is valid (if not replaced)
+        cloud_provider = Provider().getInstance(cfg.provider)
+        prices = cloud_provider.get_spot_prices(regions)
```

Comment:
You can't do this in the loop. You're requesting all of the data for previous provider/regions each time. (n! instead of n) `regions` will be invalid once we have more than one provider, since you're requesting all regions (for all providers) from each provider in use.

Reply:
Yeah, I now realized this only worked because we have only one provider. I think the new method will work much better.